ocaml/xapi/fileserver.ml

```diff
@@ -58,40 +58,44 @@ let send_file (uri_base : string) (dir : string) (req : Request.t)
     (bio : Buf_io.t) _ =
   let uri_base_len = String.length uri_base in
   let s = Buf_io.fd_of bio in
-  Buf_io.assert_buffer_empty bio ;
-  let is_external_http = is_external_http req s in
-  if is_external_http && !Xapi_globs.website_https_only then
+  if !Xapi_globs.disable_webserver then
     Http_svr.response_forbidden ~req s
-  else if is_external_http && Option.is_some req.Request.host then
-    (* Redirect towards HTTPS *)
-    let host = Option.get req.Request.host in
-    let path = req.Request.uri in
-    let dest = Uri.make ~scheme:"https" ~host ~path () |> Uri.to_string in
-    Http_svr.response_redirect ~req s dest
-  else
-    let uri = req.Request.uri in
-    try
-      let relative_url =
-        String.sub uri uri_base_len (String.length uri - uri_base_len)
-      in
-      (* file_path is the thing which should be served *)
-      let file_path = dir ^ "/" ^ relative_url in
-      (* remove any dodgy use of "." or ".." NB we don't prevent the use of symlinks *)
-      let file_path =
-        Xapi_stdext_unix.Unixext.resolve_dot_and_dotdot file_path
-      in
-      if not (String.startswith dir file_path) then (
-        debug "Rejecting request for file: %s (outside of directory %s)"
-          file_path dir ;
-        Http_svr.response_forbidden ~req s
-      ) else
-        let stat = Unix.stat file_path in
-        (* if a directory, automatically add index.html *)
+  else (
+    Buf_io.assert_buffer_empty bio ;
+    let is_external_http = is_external_http req s in
+    if is_external_http && !Xapi_globs.website_https_only then
```

Comment:
Could this expression not be extended to get the desired outcome? I don't like the deeper and deeper nesting if it could be avoided.

Reply:
As a configuration option, this is fine with me.

Reply:
I've tried using options, along with bind, to indicate whether a single reply has been sent back to the client, but it ends up more verbose, and not quite as clean as I would like:

```ocaml
let reply_maybe condition reply =
  if condition () then
    Some (reply ())
  else
    None

...

  let reply () =
    let* () =
      reply_maybe
        (fun () -> is_external_http && !Xapi_globs.website_https_only)
        (fun () -> Http_svr.response_forbidden ~req s)
    in

...

  in
  match reply () with
  | Some () ->
      ()
  | None ->
      Http_svr.response_missing s (missing uri)
```